tools/depcheck/pkg/cmd/trace/filter.go

```diff
@@ -0,0 +1,82 @@
+package trace
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gonum/graph/concrete"
+
+	depgraph "github.com/openshift/origin/tools/depcheck/pkg/graph"
+)
+
+// FilterVendorPackages receives a graph and a set of packagePrefixes contained within the graph.
+// Returns a new graph with the sub-tree for each node matching the packagePrefix collapsed into
+// just that node. Relationships among packagePrefixes are kept: edges originating from vendorRoot
+// subpackages are re-written to originate from the vendorRoot and edges terminating at vendorRoot
+// subpackages are re-written to terminate at the vendorRoot.
+func FilterVendorPackages(g *depgraph.MutableDirectedGraph, packagePrefixes []string) (*depgraph.MutableDirectedGraph, error) {
+	collapsedGraph := depgraph.NewMutableDirectedGraph(concrete.NewDirectedGraph())
+
+	// copy all nodes to new graph
+	for _, n := range g.Nodes() {
+		node, ok := n.(*depgraph.Node)
+		if !ok {
+			continue
+		}
+
+		collapsedNodeName := getFilteredNodeName(packagePrefixes, node.UniqueName)
+		collapsedGraph.AddNode(&depgraph.Node{
```

Comment:
duplicates are ignored?

Reply:
> duplicates are ignored?

I expected to see a Get followed by "if not found (nil?), add"

Reply:
> duplicates are ignored?

Don't ignore.  You'll end up stuck chasing weird bugs forever.  Return an error.  You could also check for a structured error if you like.

Reply:
@deads2k 
> duplicates are ignored?

Yeah, the MutableDirectedGraph#AddNode method ignores duplicates

> Don't ignore. You'll end up stuck chasing weird bugs forever. Return an error. You could also check for a structured error if you like.

gonum's concrete.DirectedGraph panics if I try adding a node with a duplicate ID. I could do the same on MutableDirectedGraph#AddNode and panic there if I see a node with a duplicate UniqueName.

Will handle not adding duplicates to the graph in this function